pkg/cmd/pulumi/install.go

```diff
@@ -0,0 +1,151 @@
+			ctx := commandContext()
+			displayOpts := display.Options{
+				Color: cmdutil.GetGlobalColorization(),
+			}
+
+			// Load the project
+			proj, root, err := readProject()
+			if err != nil {
+				return err
+			}
+
+			span := opentracing.SpanFromContext(ctx)
+			projinfo := &engine.Projinfo{Proj: proj, Root: root}
+			pwd, main, pctx, err := engine.ProjectInfoContext(
+				projinfo,
+				nil,
+				cmdutil.Diag(),
+				cmdutil.Diag(),
+				false,
+				span,
+				nil,
+			)
+			if err != nil {
+				return err
+			}
+
+			defer pctx.Close()
+
+			// First make sure the language plugin is present.  We need this to load the required resource plugins.
+			// TODO: we need to think about how best to version this.  For now, it always picks the latest.
+			runtime := proj.Runtime
+			lang, err := pctx.Host.LanguageRuntime(pctx.Root, pctx.Pwd, runtime.Name(), runtime.Options())
+			if err != nil {
+				return fmt.Errorf("failed to load language plugin %s: %w", runtime.Name(), err)
+			}
+
+			if err = lang.InstallDependencies(root); err != nil {
+				return fmt.Errorf("installing dependencies failed; rerun manually to try again, "+
+					"then run `pulumi up` to perform an initial deployment: %w", err)
+			}
+
+			// Compute the set of plugins the current project needs.
+			installs, err := lang.GetRequiredPlugins(plugin.ProgInfo{
+				Proj:    proj,
+				Pwd:     pwd,
+				Program: main,
+			})
+			if err != nil {
+				return err
+			}
+
+			// Now for each kind, name, version pair, download it from the release website, and install it.
+			for _, install := range installs {
+				label := fmt.Sprintf("[%s plugin %s]", install.Kind, install)
+
+				// If the plugin already exists, don't download it unless --reinstall was passed.
+				if !reinstall {
+					if install.Version != nil {
+						if workspace.HasPlugin(install) {
+							logging.V(1).Infof("%s skipping install (existing == match)", label)
+							continue
+						}
+					} else {
+						if has, _ := workspace.HasPluginGTE(install); has {
+							logging.V(1).Infof("%s skipping install (existing >= match)", label)
+							continue
+						}
+					}
+				}
+
+				cmdutil.Diag().Infoerrf(diag.Message("", "%s installing"), label)
+
+				// If we got here, actually try to do the download.
+				withProgress := func(stream io.ReadCloser, size int64) io.ReadCloser {
+					return workspace.ReadCloserProgressBar(stream, size, "Downloading plugin", displayOpts.Color)
+				}
+				retry := func(err error, attempt int, limit int, delay time.Duration) {
+					cmdutil.Diag().Warningf(
+						diag.Message("", "Error downloading plugin: %s\nWill retry in %v [%d/%d]"), err, delay, attempt, limit)
+				}
+
+				r, err := workspace.DownloadToFile(install, withProgress, retry)
+				if err != nil {
+					return fmt.Errorf("%s downloading from %s: %w", label, install.PluginDownloadURL, err)
+				}
+				defer func() { contract.IgnoreError(os.Remove(r.Name())) }()
```

Comment:
Instead of just ignoring the error, would it be worth actually logging something to the user, so they can clean up manually later?  This seems like a pretty unlikely case, so maybe it's not really worth it.

Reply:
Yup could do, just log a warning that a files been left behind.